proposals/2024/12-27_dynamic-ruleset.md

```diff
@@ -0,0 +1,197 @@
+**Please provide a summary of the new feature, redesign or refactor:**
+
+<!--
+Provide your description here.
+-->
+
+Currently, Vela currently offers several methods to filter which steps are executed:
+
+1. [`ruleset` key](https://go-vela.github.io/docs/reference/yaml/steps/#the-ruleset-key).
+2. Conditional logic with [templating directly in the Vela file](https://go-vela.github.io/docs/templates/#templating-directly-in-velayml).
+3. Conditional logic with [external templates](https://go-vela.github.io/docs/templates/).
+
+This proposal aims to introduce more dynamic conditionals, similar to those available with templates, but extend this functionality to YAML pipelines that do not use any templates.
+
+**Please briefly answer the following questions:**
+
+1. Why is this required?
+
+<!-- Answer here -->
+
+There are a wide variety of use-cases that exist for this today so we'll name a few:
+
+* filter steps based on conditions that the `ruleset doesn't support`.
+* Remove the necessity of potentially complex templates merely to circumvent the limitations of `ruleset`.
+
+2. If this is a redesign or refactor, what issues exist in the current implementation?
+
+<!-- Answer here -->
+
+N/A
+
+3. Are there any other workarounds, and if so, what are the drawbacks?
+
+<!-- Answer here -->
+
+As mentioned previously, you can use [templates](https://go-vela.github.io/docs/templates/).
+
+4. Are there any related issues? Please provide them below if any exist.
+
+<!-- Answer here -->
+
+- https://github.com/go-vela/community/issues/826
+- https://github.com/go-vela/community/issues/524
+- https://github.com/go-vela/community/issues/306
+
+## Design
+
+<!--
+This section is intended to explain the solution design for the proposal.
+
+NOTE: If there are no current plans for a solution, please leave this section blank.
+-->
+
+**Please describe your solution to the proposal. This includes, but is not limited to:**
+
+* new/updated endpoints or url paths
+* new/updated configuration variables (environment, flags, files, etc.)
+* performance and user experience tradeoffs
+* security concerns or assumptions
+* examples or (pseudo) code snippets
+
+<!-- Answer here -->
+
+### Database
+
+N/A
+
+### UI
+
+N/A
+
+### API
+
+N/A
+
+### CLI
+
+N/A
+
+### YAML
+
+A `evaluation` key will be added under `ruleset` for the new dynamic functionality.
+
+This will instruct Vela to execute the step if the `evaluation` expression returns `true`.
+
+Example pipeline:
+
+```yaml
+steps:
+  - name: simple-match
+    ruleset:
+      evaluation: "VELA_BUILD_AUTHOR == 'JordanSussman'"
+    commands:
+      - echo "this will only run if the build author name is JordanSussman"
+
+  - name: function-match
+    ruleset:
+      evaluation: "hasPrefix(VELA_BUILD_AUTHOR, 'Jordan')"
+    commands:
+      - echo "this will only run if the build author name has the prefix of Jordan"
+      
+  - name: new-and-old
+    ruleset:
+      evaluation: "VELA_BUILD_AUTHOR == 'JordanSussman'"
```

Comment:
i think these are just pseudo examples. how would we actually access and make available the build env vars?

Reply:
> how would we actually access and make available the build env vars

The [example code](https://github.com/expr-lang/expr/tree/fb6792b2486778dd8a3eb5ab2e7550f5b1dad150?tab=readme-ov-file#examples) in the expr README demonstrates how to handle the underlying code for injecting env vars. Additionally, I created a simple [Go playground](https://go.dev/play/p/6dEEUPTzK8r) to showcase a more Vela-specific implementation.

Reply:
@wass3rw3rk, just to clarify, are you saying that the built-in Vela environment variables aren't available when the rulesets are evaluated, compared to how we would write the underlying expr code to inject them?

Reply:
the code example is what i was looking for - ty! i thought you had to do more to reference a variable 👍🏼

Reply:
Still need to spend some time testing it out to confirm the environments are correctly populated, but I have the initial server code changes within https://github.com/go-vela/server/pull/1233.